api/agent/agent.go

```diff
@@ -113,13 +113,39 @@ type agent struct {
 	// TODO(reed): shoot this fucking thing
 	callOverrider CallOverrider
 
+	// Call Extension interceptor
+	extensionInterceptor ExtensionInterceptor
+
 	// additional options to configure each call
 	callOpts []CallOpt
 
 	// deferred actions to call at end of initialisation
 	onStartup []func()
 }
 
+type ExtensionInterceptor interface {
+	// BeforeCreate is triggered immediately preceding container creation, after all other
+	// configuration is in place.
+	// ContainerTask.Id() can be used to provide a stable per-container key.
+	BeforeCreate(context.Context, drivers.ContainerTask, drivers.Cookie, Call) error
```

Comment:
it would be nice to eliminate the cookie here, if we can - usage is unclear from this PR alone, just floating the thought out there. I could see a case for wanting to manipulate it though (with how it is now, which is a bit odd) it'd be ideal to be able to modify the containertask or call probably instead, similar to the other listener like things, which would hopefully cover 'configure the container this way' and 'add this info on the call', respectively (which are the only things i can think of wanting here) - oddly, all of these things are interfaces and not easily mutated from packages outside of their origin, which is unlike the other listeners: again, usage is just unclear here.

Reply:
The basic thrust of this is to be able to do things like paralleling the iofs-style additional mountpoint. the gist here is to be able to do two things:

1. mount a filesystem (tmpfs or something) into the container (maybe set an env var to point to the mount)
2. Per-invocation, shove some extension metadata into that filesystem for access from user code.

The point about this is to bolt on "automagic" delivery of creds (amongst other things - there are perhaps some other uses of this around debugging) into an invocation.

(If credentials live for less than the runtime of an instance, it may be the case in the future that we want to bolt on an "UpdateMetadata" message to the Exchange protocol - which could refresh creds and their ilk through this.)

If there's a decent way at the container def via ContainerTask I'll look to do that instead.

Reply:
(The lack of context here is a bit of a PITA, I agree. Credential management is pretty service-specific; there's not a huge amount that can be done around that I think :-( )